File: src/data/dat1.cpp
// rivet project
// Copyright (c) 2023 <https://github.com/yretenai/rivet>
// SPDX-License-Identifier: MPL-2.0

#include <memory>
#include <utility>

#include <ankerl/unordered_dense.h>

#include <rivet/data/dat1.hpp>

namespace rivet::data {
	dat1::dat1(const std::shared_ptr<rivet_data_array> &stream, const std::shared_ptr<rivet_data_array> &resident_stream): buffer(stream), resident_buffer(resident_stream) {
		if (buffer->get<uint32_t>(0) != magic) {
			throw invalid_tag_error("dat1::dat1: invalid tag");
		}

		header = buffer->get<dat1_header>(0);

		if (header.section_count > 0) {
			const auto section_headers = buffer->slice<dat1_entry>(sizeof(dat1_header), header.section_count);

			const auto resident_start = buffer->size();
			const auto resident_end = resident_start + (resident_buffer == nullptr ? 0 : resident_buffer->size());
			for (auto section_header : *section_headers) {
				section_ids.emplace(section_header.type_id);

				if (section_header.offset > resident_end) {
					throw invalid_operation("dat1::dat1: section offset is out of bounds");
				}

				if (section_header.offset < resident_start) {
					auto slice = buffer->slice(section_header.offset, section_header.size);
					sections.emplace(section_header.type_id, std::make_pair(section_header, slice));
				} else {
					auto slice = resident_buffer->slice(section_header.offset - resident_start, section_header.size);
					sections.emplace(section_header.type_id, std::make_pair(section_header, slice));
				}
			}
		}

		type_name = buffer->to_cstring_view(sizeof(dat1_header) + sizeof(dat1_entry) * header.section_count);
	}

	auto
	dat1::get_section_data(const rivet_type_id type_id) const -> std::shared_ptr<rivet_data_array> {
		const auto entry = sections.find(type_id);

		if (entry == sections.end()) {
			return nullptr;
		}

		return entry->second.second;
	}
} // namespace rivet::data

File: Rivet/Data/DAT1.cs
// SPDX-FileCopyrightText: 2025-2026 Neptuwunium
//
// SPDX-License-Identifier: EUPL-1.2

using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text;
using Pluto.Extensions;
using Pluto.IO.Binary;
using Rivet.IO;
using Rivet.Models.Data;

namespace Rivet.Data;

public sealed class DAT1 : IStringPooled {
	public const uint MagicValue = 0x44415431u;

	/// <remarks>Ownership of the buffers is NOT TRANSFERRED to this class, the calling class must dispose it.</remarks>
	public DAT1(IRentedArray<byte> buffer) : this([buffer]) { }

	/// <remarks>Ownership of the buffers is NOT TRANSFERRED to this class, the calling class must dispose it.</remarks>
	public DAT1(List<IRentedArray<byte>> buffers) {
		Buffers = buffers;

		using var reader = new ArrayPoolBinaryReader(Buffers[0], true);

		if (reader.Peek<uint>() != MagicValue) {
			throw new InvalidDataException("Invalid magic value");
		}

		Debug.Assert(Header.Reserved == 0);

		Header = reader.Read<DAT1Header>();
		var sectionHeaders = reader.Read<DAT1Entry>(Header.SectionCount);
		TypeName = reader.ReadCString<byte>(Encoding.ASCII);
		foreach (var sectionHeader in sectionHeaders) {
			var bufferOffset = 0;
			foreach (var buffer in Buffers) {
				if (sectionHeader.Offset < bufferOffset + buffer.Length) {
					Sections[sectionHeader.TypeId] = (sectionHeader, new UnownedRentedArray<byte>(buffer, sectionHeader.Offset - bufferOffset, sectionHeader.Size));
					goto nextSection;
				}

				bufferOffset += buffer.Length;
			}

			throw new InvalidOperationException("Section offset is out of bounds");

		nextSection:
			var size = sectionHeader.Offset + sectionHeader.Size;
			if (size > Size) {
				Size = size;
			}
		}
	}

	public DAT1Header Header { get; }
	public List<IRentedArray<byte>> Buffers { get; }
	public Dictionary<RivetTypeId, (DAT1Entry Entry, IRentedArray<byte> Buffer)> Sections { get; } = [];
	public string TypeName { get; }
	public int Size { get; }

	public string GetString(int offset) => Buffers[0].Span[offset..].ReadString(Encoding.ASCII) ?? string.Empty;

	public IRentedArray<byte> GetSection(ReadOnlySpan<byte> name) => GetSection(RivetTypeId.Checksum(name));
	public IRentedArray<byte> GetSection(uint hash) => !Sections.TryGetValue(hash, out var section) ? RentedArray<byte>.Empty : section.Buffer;
	public ReadOnlySpan<T> GetSection<T>(ReadOnlySpan<byte> name) where T : struct => GetSection<T>(RivetTypeId.Checksum(name));
	public ReadOnlySpan<T> GetSection<T>(uint hash) where T : struct => MemoryMarshal.Cast<byte, T>(GetSection(hash).Memory.Span);
}
